### PR TITLE
Adds /_health endpoint for the service

### PR DESCRIPTION
**Pull request by vmesel**
Adds a new endpoint for generic health checking, as requested by @cuducos.

ref #606

### CODE REVIEW

**Comment by avelino** [comment]
@vmesel check issue #606

**Review by avelino — CHANGES_REQUESTED** [review]
@vmesel check issue https://github.com/prest/prest/issues/606

**Comment by vmesel** [comment]
> @vmesel check issue #606

What should be the checks that we should perform? Now I'm just checking if the connection is available and if it's possible to query the database.

We should probably check for permissions too.

I'll need help with mocking the different database states. I'm still figuring out how to do a decent dependency injection on this code.

File: router/router.go
Comment (avelino) [inline router/router.go:36]: Change endpoint from `/health` to `/_health`, we should default to an endpoint starting with `_` for “fixed” (not dynamic) endpoints

**Comment by avelino** [comment]
@vmesel your code is great for a start, pinging the database to check for a healthy connection is enough to **health** first version.

> With time, we evolve the feature

File: controllers/generic.go
Comment (arxdsilva) [inline controllers/generic.go:1]: we should have a better name for this file, I believe `healthcheck.go` would be just fine

File: controllers/generic.go
Comment (arxdsilva) [inline controllers/generic.go:48]: A couple things:

1. This func could have a better name, `HealthStatus` would be just fine.
2. I believe the best failure response would be something in the 5xx family, I'd personally use [503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503).

File: controllers/generic_test.go
Comment (arxdsilva) [inline controllers/generic_test.go:29]: wouldnt this line be this?
```
		testutils.DoRequest(t, server.URL+tc.url, nil, tc.method, tc.status, "HealthStatus")
```

**Comment by arxdsilva** [comment]
also please dont forget to add docs about this endpoint

File: controllers/generic.go
Comment (avelino) [reply controllers/generic.go:48]: > 2. I believe the best failure response would be something in the 5xx family, I'd personally use [503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503)

`5xx` server error: this problem was not a prestd server error but a client (database) error

> `424` **Failed Dependency:**
> The 424 (Failed Dependency) status code means that the method could not be performed on the resource because the requested action depended on another action and that action failed. For example, if a command in a PROPPATCH method fails, then, at minimum, the rest of the commands will also fail with 424 (Failed Dependency).

In the [Healtch Check **RFC**](https://datatracker.ietf.org/doc/html/draft-inadarei-api-health-check-00) it describes that either `4xx or 5xx` can be used.

> *  "warn": healthy, with some concerns.
> For "pass" and "warn" statuses HTTP response code in the 2xx - 3xx range MUST be used. for "fail" status HTTP response code in the 4xx - 5xx range MUST be used.


I believe `424` fits better

**Comment by avelino** [comment]
Another important point
this endpoint must be skipped in the middleware

File: controllers/generic.go
Comment (arxdsilva) [reply controllers/generic.go:48]: > 5xx server error: this problem was not a prestd server error but a client (database) error

the problem with `4xx` client error is that this is also not a client error.

> [15.5. ](https://www.rfc-editor.org/rfc/rfc9110.html#section-15.5)[Client Error 4xx](https://www.rfc-editor.org/rfc/rfc9110.html#name-client-error-4xx)
The 4xx (Client Error) class of status code indicates that the **client seems to have erred**.

I believe that a `424` would be fit if we were calling external services that were dependencies and this request failed due to a bad body or params. With this in mind I'd sugest that we just return a flat `500` as explained bellow, since this is a internal dependency error.

> [15.6.1. ](https://www.rfc-editor.org/rfc/rfc9110.html#section-15.6.1)[500 Internal Server Error](https://www.rfc-editor.org/rfc/rfc9110.html#name-500-internal-server-error)
The 500 (Internal Server Error) status code indicates that the server encountered an unexpected condition that prevented it from fulfilling the request.

Comment (vmesel) [reply controllers/generic.go:48]: I will change the controller/function name. 

I chose the 424 status code so we could imply that this was an error occuring on pSQL, not exactly on pREST.

File: controllers/generic.go
Comment (vmesel) [reply controllers/generic.go:1]: Doing that!

File: controllers/generic_test.go
Comment (vmesel) [reply controllers/generic_test.go:29]: Yes, thanks for noticing that!

**Comment by vmesel** [comment]
> Another important point this endpoint must be skipped in the middleware

What is the best way to do this? I'm trying to add a new router on Negroni but it seems to be wrapped with the same middleware. Should I add the route before those middlewares on the list?

**Review by avelino — APPROVED** [review]

File: controllers/generic.go
Comment (vmesel) [reply controllers/generic.go]: As we spoke on twitter, I'm letting the error be a 503 so we can understand that the service is unavailable and it's not necessarily a pREST bad-functioning part.

File: controllers/healthcheck_test.go
Comment (cuducos) [inline controllers/healthcheck_test.go:19]: 1. Aren't we testing the failure?
2. If not, do we need a test cases table for a single test?

File: docs/api-reference/queries.md
Comment (cuducos) [inline docs/api-reference/queries.md:136]: How does these changes on this file relates to this PR, can you clarify?

File: docs/deployment/server-configuration.md
Comment (cuducos) [inline docs/deployment/server-configuration.md:185]: Isn't it 503?

**Review by cuducos — CHANGES_REQUESTED** [review]

File: controllers/healthcheck_test.go
Comment (vmesel) [reply controllers/healthcheck_test.go:19]: So, to test the failure I need to mock the database connection and I'm not finding a suitable way to do it. Can you help me with that?

But yes, we **are not** testing for failure

File: controllers/healthcheck.go
Comment (vmesel) [reply controllers/healthcheck.go:47]: My fault.

File: controllers/healthcheck.go
Comment (vmesel) [reply controllers/healthcheck.go:15]: This structure was created for testing purposes, but it's being used for actual code as well. I'm doing this to try to inject a dependency through the interface as shown on: https://www.thegreatcodeadventure.com/mocking-http-requests-in-golang/

File: docs/api-reference/queries.md
Comment (vmesel) [reply docs/api-reference/queries.md:136]: It's not related to this PR. But it's needed for a better comprehension of the function, I've done this while solving a work issue related to this function and forgot to remove the commit from the stash.

File: controllers/healthcheck_test.go
Comment (cuducos) [inline controllers/healthcheck_test.go:30]: Not sure if this is a pattern from pREST but any reason to not use the native `.t.Run(…)`?

```suggestion
	for _, tc := range testCases {
		t.Run(tc.description, function (t *testing.T) {
				testutils.DoRequest(t, server.URL+tc.url, nil, tc.method, tc.status, "HealthStatus")
		}
	}
```

File: controllers/healthcheck_test.go
Comment (cuducos) [reply controllers/healthcheck_test.go:19]: I cannot. I haven't dug that further on the abstraction of the `testutils.DoRequest`. What I would do if I were on your shoes:

1. take a look at other tests just in case they are testing HTTP failures somewhere else (maybe you've covered that already, np)
2. as on pREST community
3. wait or call for opinions on other reviewers here : )

File: controllers/healthcheck.go
Comment (cuducos) [reply controllers/healthcheck.go:15]: Awesome. So we need to update the docs/comment : )

File: docs/api-reference/queries.md
Comment (cuducos) [reply docs/api-reference/queries.md:136]: Ok, thanks for the clarification. I just found it confusing given the context : )

File: controllers/healthcheck.go
Comment (fabriziomello) [inline controllers/healthcheck.go:33]:
```suggestion
	_, err = db.Exec(";")
```
Using just `;` is enough for this check and less expensive because it will not require parsing/planning. Another approach is to use `db.Ping()` (not sure but I guess behind the scenes it uses PQping or PQpingParams from libpqw - https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PQPINGPARAMS) and if it is true it is even less expensive for the backend.

**Review by avelino — CHANGES_REQUESTED** [review]
```
Error: controllers/healthcheck_test.go:28:49: missing ',' in argument list
[23](https://github.com/prest/prest/actions/runs/3800347709/jobs/6463736921#step:10:24)
FAIL
```

**Comment by vmesel** [comment]
> ```
> Error: controllers/healthcheck_test.go:28:49: missing ',' in argument list
> [23](https://github.com/prest/prest/actions/runs/3800347709/jobs/6463736921#step:10:24)
> FAIL
> ```

Tks @avelino

File: controllers/healthcheck_test.go
Comment (vmesel) [reply controllers/healthcheck_test.go:19]: @avelino @arxdsilva can you guys help us here?

**Review by avelino — APPROVED** [review]
LGTM

**Comment by vmesel** [comment]
I just need to change the way I mocked the database requests to GoMock as requested by @arxdsilva. You guys can review it, but you will need to re-review it later.

I can send you guys a message on slack after it's ready.

File: Makefile
Comment (cuducos) [inline Makefile:7]: I couldn't find in the PR description a justification for that change. Can you expand on that? Also, if we're not using `DOCKER_COMPOSE` anymore, what is the reason to keep it on line 1?

File: go.mod
Comment (cuducos) [inline go.mod:8]: To understand since I couldn't find that in the PR description: is this test case so unique compared to other tests in pREST that we need to add a dependency for it, or is it a general direction in the pREST to migrate tests to use this library?

I am not a maintainer, but I think that might make the onboarding more difficult to pREST; that's why I ask. Having a standard on how to test things is important.

File: docs/deployment/server-configuration.md
Comment (cuducos) [inline docs/deployment/server-configuration.md:233]: The general use of a health check is to know that the web server is responding (not that the database is reachable). The database being reachable is a requirement of the web server.

File: controllers/healthcheck.go
Comment (cuducos) [inline controllers/healthcheck.go:61]: NIT
```suggestion
			http.Error(w, "unable to run queries on the database", http.StatusServiceUnavailable)
```

File: controllers/healthcheck_test.go
Comment (cuducos) [inline controllers/healthcheck_test.go:46]: I think I have asked that before, but… if we have only one test case, why build a test case table? Are we adding more tests?

**Review by cuducos — COMMENTED** [review]
I still don't think the changes on `docs/api-reference/queries.md` should be in this PR, since @vmesel [said it's unrelated](https://github.com/prest/prest/pull/750/files#r1058351179). On top of that, I added some questions for clarification, most impottant regarding the usage of mocks in pREST.

File: controllers/healthcheck_test.go
Comment (vmesel) [reply controllers/healthcheck_test.go:46]: There is no need for it, but, I chose to stand with it to keep the pattern observed on previous tests. I can simplify this for sure. Do you think it would be better, cuduquitos <@cuducos> ?

File: docs/deployment/server-configuration.md
Comment (vmesel) [reply docs/deployment/server-configuration.md:233]: Should the description be: `The server will return 503 whenever the web server and pREST requirements are not responding, such as Postgres connection or querying access`?

File: Makefile
Comment (vmesel) [reply Makefile:7]: No reason, don't know what happened here. Going to rollback this part of the file

File: Makefile
Comment (cuducos) [inline Makefile:11]: Is there a reason to use `$(DOCKER_COMPOSE)` on `make build_test_image`, but `docker-compose` on `test` and `test_fileoutput`?

File: Makefile
Comment (cuducos) [inline Makefile:15]: Do we really need it?
I mean… we're hardcoding a file name `testing.txt` that user might prefer to be somewhere else. Also, this file is not on `.gitignore`

Have you considered removing `test_fileoutput` and instead using `make test > MY_FILE_PATH`?

I see this way as a better option:
- we keep `.git` save by not producing files that should not be committed
- we avoid polluting `.gitignore`
- we give the user freedom to create this file else where (e.g. in `/tmp`, that would be my choice)

The only trade-off I see is that it `make test > /tmp/test-output` would include the `go generate ./...` — but overall I feel like a good trade off. What do you think?

Comment (vmesel) [reply Makefile:15]: It looks good to me!

File: Makefile
Comment (vmesel) [reply Makefile:11]: Fixing.

File: mocks/healthcheck.go
Comment (arxdsilva) [inline mocks/healthcheck.go:64]: its not a pattern but would be easier to have this file under `/controllers/mock` so we keep all controllers mock inside instead of doing one general folder for all mocks. This helps us being more objective with names, responsibilities and provides a meaningful import path.

File: go.mod
Comment (arxdsilva) [reply go.mod:8]: I'd say that it will be a tendency as we continue to develop the app to add more of this as it helps reducing the test complexity.

Also @vmesel please use the latest version, `v1.6.0`

File: controllers/healthcheck.go
Comment (arxdsilva) [inline controllers/healthcheck.go:48]: please move this closer to its usage

File: controllers/healthcheck.go
Comment (arxdsilva) [reply controllers/healthcheck.go:51]: Also if `RunTestQuery()` also uses this method and return its possible error you could do this closure there and remove this call

Comment (vmesel) [reply controllers/healthcheck.go:51]: @arxdsilva I did this with the intent of showing two possible errors: Code execution error (a querying error) and a connection error. If you think this is a deal-breaker, I can remove it.

Comment (cuducos) [reply controllers/healthcheck.go:51]: I am afraid @vmesel didn't get @arxdsilva's point — or maybe I got it wrong. Anyway… what I have in mind is that the closure and the close call would look like this:

```go
conn, err := func() {
	db, err = postgres.Get()
	if err != nil {
		return nil, err
	}
    return db, nil
}()
if err != nil {
    // …
}
defer conn.Close()
```

Which could be simplified to:

```go
conn, err := func() {
	return postgres.Get()
}()
if err != nil {
    // …
}
defer conn.Close()
```

Which is the same as:

```go
conn, err := postgres.Get()
if err != nil {
    // …
}
defer conn.Close()
```

There are two really important things going on here. I agree with one and disagree with the other.

💯 on the `defer conn.Close()` — we need that.

But working with the closure that way would make it difficult to test different scenarios using mock or dependency injection. So I'd be more careful and investigate further the testing options there.

Finally, Vini, this structure doesn't affect your objective of showing two possible errors.

File: docs/api-reference/queries.md
Comment (cuducos) [reply docs/api-reference/queries.md:136]: Just checking: you marked that as _resolved_, but you keep unrelated PR changes. Is that correct? If so, what is the reasoning?

Comment (cuducos) [reply docs/api-reference/queries.md:136]: The only scenario in which this would make sense is the one in which these changes, even if unrelated, are needed.

If that's the case, you should open another branch and PR for these changes. Since we raised that 2 weeks ago, if these changes are needed, and if they were in a dedicated PR, they would probably be merged already by now.

File: controllers/healthcheck.go
Comment (vmesel) [reply controllers/healthcheck.go]: I think I was in need to take a cup of caffeine. Thank you for the explanation, @cuducos. It doesn't mess up, I was thinking of single error types for the same function but this does not need to be true. Now I got it

Comment (arxdsilva) [reply controllers/healthcheck.go]: @vmesel @cuducos I believe that we can further simplify those function to something like this:

```go
func (d DBConn) RunTestQuery() (err error) {
	db, err := d.GetConnection()
	if err != nil {
		return err
	}
	defer db.Close() // please check if exists

	_, err = db.Exec(";")
	return err
}

func WrappedHealthCheck(dbc DbConnection) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		// this is abstracted and much more testable/mockable 
		// since it doesnt know any db logic
		err := dbc.RunTestQuery() 
		if err != nil {
			http.Error(w, "unable to run queries on the database", http.StatusServiceUnavailable)
			return
		}

		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(HealthCheck{
			Status: "ok",
		})
	}
}

```

OBS: added comments just to explain more, but this small change should fix the issue, maybe even we could on another step think about making this getConnection private, but I dont mind to keep it that way.

**Comment by avelino** [comment]
I will share what i think about **gomock** (`github.com/golang/mock`).

I didn't follow the RFC that gave as initiative the development of it, but personally, I didn't like the team working with mock using it.

there is probably a super logical explanation for it being the way it is that I don't know about (lacking searching for information).

> just sharing what I talked to @vmesel in private
> cc @arxdsilva

**Comment by vmesel** [comment]
@avelino @arxdsilva I've sat down with @cuducos this afternoon to finally review and fix anything major that appeared as a problem. Could you guys review it for us, so we could proceed?

**Review by arxdsilva — APPROVED** [review]
Thanks for your collaboration @vmesel ! Again sorry for the long discussions.

**Comment by arxdsilva** [comment]
@vmesel there are some conflicts in the branch, could you please rebase and push again so I can merge?

**Comment by avelino** [comment]
@arxdsilva after this merge deserves a release with all the issues that was implemented later in mind, can you generate it?

> It is worth fixing this #734 first, it is probably docker hub tokens.

**Pull request merged**
